✨ v1: 기본적 단축 url 서비스 구현 / v2 : 첫번째 리팩토링

healtheefart opened the pull request:
## 💻 작업 내역

- 기본적인 단축 url 서비스(v1)를 만들었습니다.

## 🔎 PR 특이 사항 

- **HomeController/getHome()** : return 값을 "home"으로 하여 template/home.html로 로드되게 하려했는데 원하는 대로 나오지 않아(home.html이 아닌 home이라고 적힌 페이지가 로드됨) ModelAndView로 return했습니다.
- **UrlShortenerController** : originUrl로 redirect할 때 위와 같은 방법으로 string을 return하여 리다이렉트하려했는데 실패하여 httpServletResponse를 통한 redirect 방법을 사용했습니다.

## 🛠️ refactoring list

- url validation
- shorten-url의 길이를 적절하게 적용하기 위해 BigInteger id로 변경
- exception
- yml 민감정보 숨기기

healtheefart committed:
chore: change properties into yml

healtheefart committed:
chore: added .gitignore

healtheefart committed:
chore: added dependency of mysql, thymeleaf, lombok, jpa

healtheefart committed:
Feat : created BASE62Utils
- (encode) : url의 id를 encode한 값 return
- (decode) : encode한 값을 decode한 url의 id return

healtheefart committed:
Feat : created controller
- (HomeController) : 홈화면, redirect
- (UrlShortenerController) : url 단축

healtheefart committed:
Feat : created UrlShortenerService

healtheefart committed:
Feat : created UrlShortenerRepository

healtheefart committed:
Feat : created ShortenUrl(entity), ShortenUrlRequest(dto), ShortenUrlResponse(dto)

healtheefart committed:
Feat : created home.html(thymeleaf page), style.css

sunH0 commented on src/main/java/com/mission/ShortURLService/utils/BASE62Utils.java:
```diff
@@ -0,0 +1,32 @@
+		} while (value > 0);
+		return sb.toString();
+	}
+	public static Integer decode(String value) {
```
인코딩된 url을 db에 저장하지 않고 디코딩을 하는 방법을 선택하신 이유가 있으실까요?

sunH0 replied:
저는 DB 메모리 공간효율을 높일 수  있고 실제로 decode 연산하는 시간이 오래 걸리진 않을 것 같아서 좋다고 생각합니다!

healtheefart replied:
아 이 부분은 [요 부분](https://github.com/study-for-Coding-Mission/Short-URL-Service/pull/10#discussion_r885885869) 에서 문제가 생겨가지고 직접 decode하는 식으로 구현했습니다... 다시 한번 setter로 db에 적용이 되는지 시도해보고 적용이 된다면 수정할 부분입니다!

sunH0 replied:
db를 활용하지 않고 decode 해주는 것도 좋은 방법이라 생각해요! 하지만 set에서 생긴 문제에 대한 원인은 알아보면 좋을 것 같습니다~

healtheefart replied:
알고보니 제가 서비스에 트랜잭션을 걸어두지 않아서 생긴 해프닝이었습니다 하핫

sunH0 replied:
트랜젝션을 설정하면 메소드 전체를 하나의 작업으로 묶을 수 있어서 1차 캐시를 활용할 수 있습니다!

sunH0 commented on src/main/java/com/mission/ShortURLService/service/UrlShortenerService.java:
```diff
@@ -0,0 +1,37 @@
+		if (optional.isEmpty()) {
+			throw new RuntimeException();
+		}
+		return optional.orElseThrow(() -> new RuntimeException()).getOriginUrl();
```
위에 if문과 중복되지 않나요??

healtheefart replied:
앗ㅋㅋㅋㅋㅋㅋ수정하겠습니다!!

healtheefart replied:
[7087150](https://github.com/study-for-Coding-Mission/Short-URL-Service/pull/7/commits/708715019cc848eb60081e6da4b80815c778d1bb)로 수정했습니다!

sunH0 replied:
확인했습니다! 가독성이 좋도록 orElseThrow 코드는 find 메소드에 작성하는 것도 좋을 것 같습니다~

sunH0 reviewed (commented):
수고하셨습니다! yml 민감정보 숨기기는 저도 궁금하네요!

sunH0 commented on src/main/resources/templates/home.html:
```diff
@@ -0,0 +1,47 @@
+      <input value="SHORTEN!" type="button" onclick="getResult()">
+    </div>
+
+    <div id="result">
```
혹시 단축 결과는 어떻게 보여주나요??

sunH0 commented on src/main/resources/templates/home.html:
```diff
@@ -0,0 +1,47 @@
+    }).done(function(result) {
+      let parse = result;
+      console.log(parse);
+      let response = "<p>shorten url: <a href=" + parse +" target='_blank'>" + parse +"</a></p>";
```
단축 결과가 밑에 텍스트로 뜨는 건가요??

healtheefart replied:
헉 이 부분은 제가 custom한 ApiResponse로 감싼 응답값은 화면에 출력이 안 되길래 ResponseEntity로 바꾸면서 시험해본다고 바꿨다가 다시 ctrl+z를 안 했나봐요  하핫
앞으로 commit하기 전에 좀 더 꼼꼼히 보고 pr 날릴게유~

healtheefart replied:
[a636737](https://github.com/study-for-Coding-Mission/Short-URL-Service/pull/7/commits/a636737cc2d3cbe2b5b8bc6c19b8889db576a2d0) 로 수정했습니다!

sunH0 replied:
아하 html 코드에서 단축된 url 을 보여주는 box 가 안보여서 궁금했어요

sunH0 commented on src/main/java/com/mission/ShortURLService/controller/HomeController.java:
```diff
@@ -0,0 +1,31 @@
+	private final UrlShortenerService urlShortenerService;
+
+	@GetMapping
+	public ModelAndView getHome() {
```
WebMvcConfigurer 의 addViewControllers 를 사용하는 것도 좋을 것 같아요!

healtheefart replied:
오호라 넵! spring mvc에 대해서 좀 더 알아보고 수정하겠습니다!

sunH0 commented on src/main/java/com/mission/ShortURLService/controller/HomeController.java:
```diff
@@ -0,0 +1,31 @@
+	public void redirect(@PathVariable(value = "encoded-url") String encodedUrl, HttpServletResponse httpServletResponse)
+		throws IOException {
+		String originUrl = urlShortenerService.getOriginUrl(encodedUrl);
+		httpServletResponse.sendRedirect("http://" + originUrl);
```
http:// 가 중복될 수도 있을 것 같아요!

healtheefart replied:
추후에 refactoring하면서 http://를 포함한 input과 없는 input에 대해 모두 고려하도록 앞에 http://가 없는 input은 추가하는 로직을 구현해볼 생각입니다!

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/ShortenUrlResponse.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+public class ShortenUrlResponse {
+	private String originUrl;
```
originUrl 까지 반환하는 이유가 있을까요?

healtheefart replied:
```javascript
let parse = JSON.parse(result);
let response = "<p>shorten url: <a href=" + "http://" + parse.originUrl +" target='_blank'>" + parse.shortenUrl +"</a></p>";
```
home.html에서 응답값을 받아와서 표시해주는 부분의 코드인데요
a태그의 href를 통해서도 바로 redirect할 수 있도록 굳이 넣어봤습니다!
근데 지금 생각해보니 그렇게 되면 responseDto가 entity에서 id를 뺀 전체 데이터가 다 들어있는 dto가 돼서 dto를 따로 만든 의미가 있나싶기도 하네요...?

healtheefart replied:
선호님 생각은 어떠셔요?

sunH0 replied:
entity와 관계없이 dto는 서비스와 controller 계층 사이에서 DB를 보호하기 위해 필요하다 생각해요!
redirect는 단축된 url 을 통해서 동작되니까 원본 url은 반환할 필요가 없지 않을까요??

healtheefart replied:
맞아요 처음에는 해당 url로 바로 리다이렉트할 수 있도록 response에 originUrl이 필요하다고 생각했었는데 바로 리다이렉트하는 부분은 뷰에서도 접근할 수 있는 부분($("#originUrl").val())이라 굳이 필요없더라고요..!
그래서 어제 리팩토링 작업 중에 수정하였습니다!

healtheefart committed:
refactor : text에서 JSON형식으로 parse한 응답값으로 반환하도록 수정

healtheefart committed:
refactor : getOriginUrl 중복된 코드 수정

healtheefart committed:
chore : added log4j, springboot validation dependency

healtheefart committed:
refactor : WebConfigurer를 통한 index.html 뷰 반환
- index.html : renamed from home.html
- RedirectController.java
  - renamed from HomeController.java
  - removed getHome()
- ShortenUrlResponse : removed originUrl field

healtheefart committed:
naming & refactor : ShortenUrl.java -> Url.java로 수정
- ShortenUrlRequest와 네이밍을 구별하기 위해 엔티티를 Url.java로 네이밍 수정
- Url Id를 BigInteger로, 시작Value를 100000으로 주어 encode값을 일정 길이 이상이 되게끔 수정

healtheefart committed:
naming & refactor : ShortenUrlDto -> UrlShortenerDto로 수정
- ShortenUrlRequest.java -> UrlShortenerRequest.java
- ShortenUrlResponse.java -> UrlShortenerResponse.java
- added @Validated to request.originUrl

healtheefart committed:
refactor : util관련 값을 설정파일로 관리

healtheefart committed:
refactor : request값 @Validated 추가

healtheefart committed:
refactor : UrlShortenerService.java 전면적 refactoring
- added @Transactional annotation
- createNewUrl()을 통해 새로 url 단축로직 분리

sunH0 commented on src/main/java/com/mission/ShortURLService/utils/BASE62Utils.java:
```diff
@@ -0,0 +1,35 @@
+
+	public static String encode(String value) {
+		StringBuilder sb = new StringBuilder(HEAD);
+		Integer intVal = Integer.valueOf(value);
```
결국 여기서 정수형으로 변환을 하게 되는데 String으로 value를 받을 이유가 있을까요? 
UrlShortenerService  에서도 id를 toString 해서 넘기는 것으로 확인이 되네요!

그리고 int가 아닌 Integer를 사용하신 이유가 궁금합니다!

healtheefart replied:
1) String으로 받은 이유 : 
- 일단 처음 id를 BigInteger로 주게된 이유는 encode되는 값이 한 글자여서 좀 더 길이를 늘리고자 initialValue를 높게 주고, 사용할 수 있는 id가 줄어들 것을 감안해 int값이 아닌 BigInteger로 주었습니다.
- 링크에 대한 id의 자료형이 BigInteger이다보니 int형으로 변환하면서 데이터 손실이 없도록 하기 위해 String으로 받고, int 범위를 넘어가면 String을 잘라서 사용하려 했는데요. 그 부분 구현보다 먼저 테스트를 통해 encode된 값이 몇 자리가 나오는지 확인해보았는데 해당 단축 알고리즘으로는 100000 이상으로 initialValue로 주어도 두 글자밖에 나오지 않아 고민이 됐습니다.
- 그래서 단축 알고리즘을 변경하거나, id의 값 자체를 높게 주는 것이 아닌 encode되는 과정에서만 id에 맞는 높은 값을 만들어보는 등의 방법을 고민해보고 있습니다!

2) Integer를 사용한 이유 : intellij에서 제공하는 introduce local variable 자동완성 기능을 사용하다보니 그렇게 됐어요!ㅎㅎ
생각해보니 연산하기위해 자동언박싱을 계속하는 꼴이라 parseInt를 통해 int로 받아서 언박싱, 박싱 과정이 없도록 구현하는 게 맞을 것 같아요!
마지막 리팩토링으로 수정해보겠습니다!

sunH0 replied:
아하 이런 이유가 있으셨네요! BigInteger 로 하신 것은 정말 세심하게 코딩하신 것 같아요!

sunH0 commented on src/main/java/com/mission/ShortURLService/utils/BASE62Utils.java:
```diff
@@ -0,0 +1,35 @@
+
+	public static String encode(String value) {
+		StringBuilder sb = new StringBuilder(HEAD);
+		Integer intVal = Integer.valueOf(value);
```
결국 여기서 정수형으로 변환을 하게 되는데 String으로 value를 받을 이유가 있을까요? 
UrlShortenerService  에서도 id를 toString 해서 넘기는 것으로 확인이 되네요!

그리고 int가 아닌 Integer를 사용하신 이유가 궁금합니다!

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/UrlShortenerRequest.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.validation.annotation.Validated;
+
+@Getter
+@Validated
```
혹시 @Validated 가 없으면 오류가 날까요??
검색해보니 서비스나 레포지토리 계층에서 사용되는 것이라고 하네요!

healtheefart replied:
에고? dto에 제가 파라미터 값에 대한 어노테이션을 달면서 같이 추가해버렸나봐요ㅠ 발견해주셔서 감사합니다!
원래 @Validated 어노테이션을 Controller에서 파라미터로 받는 dto 앞에만 추가할 생각이었습니다
그리고 제가 찾아본 자료에서는

> 파라미터의 유효성 검증은 컨트롤러에서 처리하고 서비스나 레포지토리 계층에서는 유효성 검증을 하지 않는 것이 바람직하다.
> 출처: [MangKyu's Diary:티스토리](https://mangkyu.tistory.com/174)

라고 나와 controller에서만 @Validated 어노테이션을 주어 검증해주려고 했습니다!

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/UrlShortenerRequest.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.validation.annotation.Validated;
+
+@Getter
+@Validated
```
혹시 @Validated 가 없으면 오류가 날까요??
@Valid와 어떤 차이가 있는지 궁금하네요!

healtheefart replied:
자바표준스택인 @Valid와 spring에서 제공해주는 @Validated의 차이점은 던져주는 에러도 다르기도 하지만 무엇보다 @Validated를 사용하면 유효성 검증 그룹을 커스터마이징할 수 있다고 알고 있습니다.
그래서 지금은 url 관련한 검증밖에는 없지만 마지막 리팩토링을 통해 직접 한번 사용해보려고 합니다!

sunH0 replied:
오호 @Validated에 대해서 하정님 덕분에 처음 알게 됬는데 공부해봐야 겠네요!

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/UrlShortenerRequest.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.validation.annotation.Validated;
+
+@Getter
+@Validated
```
혹시 @Validated 가 없으면 오류가 날까요??
검색해보니 서비스나 레포지토리 계층에서 사용되는 것이라고 하네요!

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/UrlShortenerRequest.java:
```diff
@@ -0,0 +1,14 @@
+	@URL
+	@NotBlank
+	private String originUrl;
+}
```
Request로 사용하는 DTO는 생성자가 없어도 되는군요! 👍
DTO에 대해 좀 더 공부해야 겠네요

healtheefart replied:
엇 제가 안 쓴 줄 몰랐는데요...하하
찾아보니 생성자가 없어도 되는 이유가
> ObjectMapper는 @RequestBody가 Property로 구현되어 있거나 생성을 위임한 경우가 아니라면 기본 생성자로 생성한다. [참고](https://velog.io/@conatuseus/RequestBody%EC%97%90-%EC%99%9C-%EA%B8%B0%EB%B3%B8-%EC%83%9D%EC%84%B1%EC%9E%90%EB%8A%94-%ED%95%84%EC%9A%94%ED%95%98%EA%B3%A0-Setter%EB%8A%94-%ED%95%84%EC%9A%94-%EC%97%86%EC%9D%84%EA%B9%8C-3-idnrafiw)

아니면
request의 필드값이 하나밖에 없어 "text로 들어온 request 자체 = 역직렬화한 dto"라 역직렬화할 필요가 없어서 일수도 있을 거 같은데 정확하게는 잘 모르겠습니다ㅠ
- [참고](https://bbbicb.tistory.com/46)

혹시 선호님은 어떻게 생각하시나요?ㅎㅎ

sunH0 replied:
우선 역직렬화 자체는 json을 java 객체로 매핑해 주는 것이기 때문에 반드시 일어난다고 생각해요. Dto에 대해서 지금까지 잘 모르고 사용했었네요... 오늘 정리해서 블로그 포스팅 해봐야 겠어요! 나중에 공유하겠습니다`

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/UrlShortenerRequest.java:
```diff
@@ -0,0 +1,14 @@
+	@URL
+	@NotBlank
+	private String originUrl;
+}
```
Request로 사용하는 DTO는 생성자가 없어도 되는군요! 👍
DTO에 대해 좀 더 공부해야 겠네요

sunH0 replied:
Request로 사용하는 DTO는 생성자가 없어도 되는군요! 👍
DTO에 대해 좀 더 공부해야 겠네요

sunH0 commented on src/main/java/com/mission/ShortURLService/controller/RedirectController.java:
```diff
@@ -0,0 +1,25 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/redirect")
+public class RedirectController {
```
Controller를 2개로 나누셨군요!

healtheefart replied:
아 넵!
처음에 하나의 controller로 구현해보았을 때에
"해당서비스url/"로 GET 요청이 왔을 때 index.html로 가는 **getHome() 메서드**가 아닌 @RequestMapping("/{encoded-url}")인 **redirect 메서드** 먼저 인식을 하고 'encoded-url'이 null값이라는 에러가 뜨더라고요...
그래서 request mapping에 우선순위를 주는 방법을 알아보았는데 찾지를 못해서 controller도 역할에 따라 나누어서 url도 다르게 구현해보았습니다.

sunH0 replied:
addViewControllers 메소드에서 특정 뷰로 요청이 갈텐데 왜 그러까요..?

healtheefart replied:
request mapping을 좀 더 알아봐야할까요...?하하 ㅠㅠ

sunH0 commented on src/main/java/com/mission/ShortURLService/controller/RedirectController.java:
```diff
@@ -0,0 +1,25 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/redirect")
+public class RedirectController {
```
Controller를 2개로 나누셨군요!

sunH0 replied:
Controller를 2개로 나누셨군요!

sunH0 commented on src/main/java/com/mission/ShortURLService/entity/Url.java:
```diff
@@ -0,0 +1,36 @@
+public class Url {
+
+	@Id
+	@SequenceGenerator(name="seq_generator", sequenceName = "url_seq", initialValue = 100000)
```
👍👍

sunH0 replied:
👍👍

sunH0 replied:
👍👍

sunH0 reviewed (commented):
리펙토링 하시느라 수고하셨습니다~
리뷰한 내용 확인 부탁드릴게요

healtheefart commented:
꼼꼼한 리뷰 감사감사 따봉 드립니다! 머지하고 마지막 리팩토링으로 돌아올게요~
